src/Container.php

```diff
@@ -161,50 +162,38 @@ private function resolveParams(string $qn, ReflectionMethod $constructor): itera
                 yield $param->getDefaultValue();
                 continue;
             }
+
             $paramType = $param->getType();
+
             if ($paramType instanceof ReflectionUnionType) {
-                $msgFormat = 'Unable to resolve %s constructor parameter $%s (position %d): it has union type %s.';
-                $types = [];
-                foreach ($paramType->getTypes() as $namedType) {
-                    $types[] = $namedType->getName();
-                }
-                $message = sprintf(
-                    $msgFormat,
-                    $qn,
-                    $param->getName(),
-                    $param->getPosition() + 1,
-                    implode('|', $types)
-                );
-                throw new ContainerException($this->buildExceptionReport($message));
+                $this->throwNonResolvableUnionParameterTypeException($qn, $param);
             }
 
-            try {
-                /** @var ReflectionNamedType $paramType */
-                $paramClass = !$paramType || $paramType->isBuiltin()
-                    ? null
-                    : new ReflectionClass($paramType->getName());
-            } catch (ReflectionException $e) {
-                $msg = sprintf(
-                    'Unable to resolve %s constructor parameter $%s: %s.',
-                    $qn,
-                    $param->getName(),
-                    $e->getMessage()
-                );
-                throw new ContainerException($this->buildExceptionReport($msg), 0, $e);
+            if ($paramType === null) {
+                $this->throwNonResolvableParameterTypeException($qn, $param, 'null provided as parameter type');
             }
 
-            if (!$paramClass) {
+            assert($paramType instanceof ReflectionNamedType);
+
+            if ($paramType->isBuiltin()) {
```

Comment:
Same as above. We can create something like $this->checkForBuiltinType() that will check for a built-in type and throw an exception.